File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
*/

class Solution {
public:

    int maxDepth(TreeNode* root) {
        return returnDepth(root);
    }

    int returnDepth(TreeNode* recievedRoot){
        if(recievedRoot == nullptr) return 0;
        int leftSideDepth  = returnDepth(recievedRoot->left);
        int rightSideDepth = returnDepth(recievedRoot->right);
        return max(leftSideDepth,rightSideDepth) + 1; // +1 for root node
    }
};

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        maxDepth = [0]
        depth = 0
        self.traverse(root,depth,maxDepth)
        return maxDepth[0]
    
    def traverse(self,root,depth,maxDepth):
        if root == None:
            return
        depth+=1
        maxDepth[0] = max(depth,maxDepth[0])
        self.traverse(root.left,depth,maxDepth)
        self.traverse(root.right,depth,maxDepth)
        depth-=1
        
        
